distributed/tests/test_scheduler.py

```diff
@@ -4890,3 +4890,93 @@ async def test_resubmit_different_task_same_key_warns_only_once(
 
     async with Worker(s.address):
         assert await c.gather(zs) == [2, 3, 4]  # Kept old ys
+
+
+def block(x, in_event, block_event):
+    in_event.set()
+    block_event.wait()
+    return x
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1, {"resources": {"a": 1}})],
+    config={"distributed.scheduler.allowed-failures": 1},
+)
+async def test_fan_out_pattern_deadlock(c, s, a):
+    """Regression test for https://github.com/dask/distributed/issues/8548
+
+    This test heavily uses resources to force scheduling decisions.
+    """
+    in_ancestor = Event()
+    block_ancestor = Event()
+    in_on_a_descendant = Event()
+    in_on_b_descendant = Event()
+    block_on_a_descendant = Event()
+    block_on_b_descendant = Event()
+
+    # Input task to 'g' that we can fail
+    with dask.annotate(resources={"b": 1}):
+        f = delayed(block)(1, in_ancestor, block_ancestor, dask_key_name="f")
+        g = delayed(inc)(f, dask_key_name="g")
+
+        # Fan-out from 'g' and run h1 and h2 on different workers
+        h1 = delayed(block)(
+            g, in_on_b_descendant, block_on_b_descendant, dask_key_name="h1"
+        )
+    with dask.annotate(resources={"a": 1}):
+        h2 = delayed(block)(
+            g, in_on_a_descendant, block_on_a_descendant, dask_key_name="h2"
+        )
+    del g
+
+    f, h1, h2 = c.compute([f, h1, h2])
+    with captured_logger("distributed.scheduler", level=logging.ERROR) as logger:
+        async with Worker(s.address, nthreads=1, resources={"b": 1}) as b:
+            await block_ancestor.set()
+            await asyncio.gather(
+                in_on_a_descendant.wait(),
+                in_on_b_descendant.wait(),
+            )
```

Comment:
```suggestion
            await in_on_a_descendant.wait()
            await in_on_b_descendant.wait()
```
nit: 5% more readable